Global and per-source gpgcheck flags

ydirson opened the pull request:
Those two `gpgcheck` and `repo-gpgcheck` flags, both declined in boot-parameter variant (`--no-gpgcheck`) and answerfile variant (`<installation gpgcheck="false">`), are meant as a replacement for the single `<installation netinstall-gpg-check>` parameter in xcp-ng 8.2.1, which was part of c86d2ac1db8a15d537fb9cea9e629a868d318ecf.

- more accurate naming than the original (was never restricted to netinstall)
- more fine-grained control, no need to remove checking of rpms when you just add new or update signed RPMs into the repo

ydirson committed:
Add support for globally disabling repo_gpgcheck

While developing and/or installing from an already-verified ISO, yum's
repo_gpgcheck just gets in the way.  This provides a general kill-switch
for checking repomd.xml.asc, through commandline or answerfile.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Add support for globally disabling gpgcheck

Similar to no-repo-gpgcheck but for RPM sigs.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented:
> more accurate naming than the original (was never restricted to netinstall)

I don't think this is true. In 8.2, we only checked signatures in the context of netinstall, AFAIK.

Wescoeur commented on answerfile.py:
```diff
@@ -101,6 +101,8 @@ def processAnswerfile(self):
                 raise AnswerfileException("Unknown mode, %s" % install_type)
 
+            results['gpgcheck'] = getBoolAttribute(self.top_node, ['gpgcheck'], default = True)
+            results['repo-gpgcheck'] = getBoolAttribute(self.top_node, ['repo-gpgcheck'], default = True)
```
Regarding the coding style, spaces are not used to reference a named param. Here `default = True`. ;)

ydirson replied:
Right, I should not have kept that space from the 8.2.1 code  ;)

Wescoeur commented on backend.py:
```diff
@@ -426,6 +426,13 @@ def add_repos(main_repositories, update_repositories, repos):
     if not main_repositories or main_repositories[0].identifier() != MAIN_REPOSITORY_NAME:
         raise RuntimeError("No main repository found")
 
+    if 'gpgcheck' in answers and not answers['gpgcheck']:
```
Why not directly `not answers.get('gpgcheck')`?

ydirson replied:
Just hesitation on my side - you're right it's just better

ydirson replied:
In fact there was a reason: `get()` returns `None` when the parameter was never specified, which would be taken as `False`, whereas the default we want is `True`.  Will add a comment to make this more explicit.

stormi replied:
I suppose we could use `get(..., True)`.

ydirson replied:
right, that's better

Wescoeur commented on backend.py:
```diff
@@ -426,6 +426,13 @@ def add_repos(main_repositories, update_repositories, repos):
+    if 'gpgcheck' in answers and not answers['gpgcheck']:
+        for repo in main_repositories: # update_repositories currently do not check anyway
+            repo.disableGpgCheck()
+    if 'repo-gpgcheck' in answers and not answers['repo-gpgcheck']:
```
Same here.

Wescoeur commented on doc/answerfile.txt:
```diff
@@ -33,6 +33,28 @@ Restore:
+  repo-gpgcheck="false"
+
+    Disable check of repodata signature, for all yum repositories
+    (`repo_gpgcheck=0` in `yum.conf`).  Don't use this for a network
```
Double spaces. ;)

ydirson replied:
Double spaces after full stop are a reflex here - OK, not in this file :disappointed:

Wescoeur commented on install.py:
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
+        # XCP-ng: no-repo-gpgcheck
+        elif opt == "--no-repo-gpgcheck":
+            results['repo-gpgcheck'] = False
+            logger.log("Commandline: no gpg check for yum repository.")
```
Well, `Commandline:` prefix is not used in the other cases. It doesn't seem very symmetrical. :slightly_smiling_face:

ydirson replied:
Right, I felt that the other messages were not quite informative enough.  In fact the only upstream log on a boot parameter is `Keymap specified on command-line`, so for consistency that should rather be something like `Yum repository gpg check disabled on command-line`  (and we should improve the messages in our other patches).

Wescoeur commented on repository.py:
```diff
@@ -254,6 +254,8 @@ def __init__(self, accessor):
         self._identifier = MAIN_REPOSITORY_NAME
         self.keyfiles = []
+        self.__gpg_check = True
+        self.__repo_gpg_check = True
```
Seems to me the convention in this project is to use a single "_" for private attributes.

ydirson replied:
I agree we should use a single underscore here, but because they are not "attributes that you do not want subclasses to use" (PEP8 wording)

Wescoeur commented on repository.py:
```diff
@@ -324,10 +326,12 @@ def _repo_config(self):
 gpgkey=file://%s
-""" % (key_path)
+""" % (1 if self.__gpg_check else 0,
+       1 if self.__repo_gpg_check else 0,
```
Can be simplified using `int(self.__gpg_check)`.

ydirson replied:
right

ydirson commented:
> > more accurate naming than the original (was never restricted to netinstall)
> 
> I don't think this is true. In 8.2, we only checked signatures in the context of netinstall, AFAIK.

I could see no explicit restriction in the original code, AFAICT I see no reason why it would not impact installs from local media.

stormi commented on install.py:
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
         elif opt == "--netinstall":
             results['netinstall'] = True
             logger.log("This is a netinstall.")
+        # XCP-ng: no-repo-gpgcheck
```
I would remove this comment and the next one, especially if we plan to contribute this upstream (without the comments) and then merge upstream back to our fork.

stormi commented on install.py:
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
             results['netinstall'] = True
             logger.log("This is a netinstall.")
+        # XCP-ng: no-repo-gpgcheck
+        elif opt == "--no-repo-gpgcheck":
```
The boot parameters (and possibly the options?) should be documented in `doc/parameters.txt`

ydirson replied:
Added in new revision

stormi replied:
I mentioned the boot parameters in priority, rather that the options handled by init.py, as that's the most useful for users. But I'm not sure what the logic is to decide what options gets described in the boot parameters and what options are described in the script parameters.

ydirson replied:
It looks like there is all installer-specific option are described under "Installer", though the "Startup" section is a mix of options interpreted by kernel and scripts launched directly by `init` (I mean `systemd` here, not the installer's own `init` script :dizzy_face:)

ydirson committed:
answerfile: support *gpgcheck override for each <source>

Global *gpgcheck flag can change the default from True to False, and these
new flags allow to override this default on a per-source basis.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented:
Are the last pushes since my approval just a rebase, or are there other changes?

ydirson commented:
> Are the last pushes since my approval just a rebase, or are there other changes?

There is a bit more, especially rework of the `backend.py` changes to prepare better for upcoming patches

stormi commented on backend.py:
```diff
@@ -402,6 +402,13 @@ def add_repos(main_repositories, update_repositories, repos):
+                        else i['gpgcheck'])
+            for repo in repos:
+                repo.setRepoGpgCheck(repo_gpgcheck)
+                repo.setGpgCheck(gpgcheck)
```
This looks like a partial stepback from what was done in the first two commits. Would it make sense to change the first two commits direclty so that the third one doesn't have to move the code around?

ydirson replied:
Yes it does.

Also, moving the code from a loop on `main_repositories` into a loop over all sources, as currently done, has the side effect of applying the `*gpgcheck` parameters to all sources, which is likely to cause issues when we suddenly perform the check on an update repo (since XenServer does not activate this check on those).

stormi commented on backend.py:
```diff
@@ -403,8 +403,10 @@ def add_repos(main_repositories, update_repositories, repos):
+            gpgcheck = (answers.get('gpgcheck', True) if i['gpgcheck'] is None
+                        else i['gpgcheck'])
             for repo in repos:
                 if repo in main_repositories:
```
Are all sources mentioned in the answerfile as a `<source>` in `main_repositories`?

ydirson replied:
Assumption was indeed that `main_repositories` comes from `<source>`'s and `update_repositories` from `<driver-source>`'s.  But that's wrong, as supplemental packs can be specified as `<source>` (as documented in `answerfile.txt`).

In more details:

## The workflow goes:
1. parsing
  * `Answerfile.parseSource()`: `<source>` -> `answers['sources']` (named `results` during parsing, `answers` starting from `backend.performInstallation()`)
  * `Answerfile.parseDriverSource()`: `<driver-source>` -> `answers['extra-repos']`
1. (bis) TUI
  * "load driver" screen adds to `answers['extra-repos']` (and does install packages inside initramfs in a non-obvious process, but that's another story, see `driver.doInteractiveLoadDriver()`)
  * the "supplemental packs installation" step at the end of install process XXXX
3. `Repository` objects creation
  * everyone goes through `repository.repositoriesFromDefinition()` and `add_repos()` regardless of "sources" vs. "extra-repos"
4. `add_repos()` separates `UpdateYumRepository`'s (i.e. those with `update.xml`) from the rest, into resp.  `update_repositories`  and `main_repositories`

Essential info: supplemental packs (which include driver packs) all include an `update.xml` (as generated by the DDK's `build-update` script).

So at this point:
- we know the install media `<source>` has no `update.xml`, so it is classified as `main_repository`
- all supplemental packs, whether from `<driver-source>` or from `<source>`, are classified as `update_repository

## What this means for this patch series:
- no supplemental pack is gpgcheck'd as per upstream code (even though some do include a sig! (e.g. `driver-microsemi-aacraid`)
- `<driver-source>` is supposed to point to supplemental packs only, so not having `*gpgcheck` attributes is not a problem until we want to allow checking sigs for them
- a `<source>` element can point declare a source with both "main repos" and supplemental packs; the latter will just ignore the `*gpgcheck` attributes.  This can probably be the case if we build an install media containing supplemental packs alongside the main repo.

## Bottom line:

* we could make the doc a little more clear about supplemental packs not honoring those attributes
* we probably want to add gpgcheck support to supplemental packs too!

stormi replied:
Thanks for the detailed explanation.

ydirson replied:
Updated the doc to document the behaviour wrt. Supplemental Packs.  Maybe we'll want to find a wording to make it sound less scary that Supplemental Packs are not gpgcheck'd ?

Wescoeur commented on answerfile.py:
```diff
@@ -267,7 +269,21 @@ def parseSource(self):
+            # handle the default situation where the attribute is not
+            # specified
+            repo_gpgcheck = (None if getStrAttribute(i, ['repo-gpgcheck'], default=None) is None
+                             else getBoolAttribute(i, ['repo-gpgcheck']))
```
Why not:
```python
getStrAttribute(i, ['repo-gpgcheck'], default=None) and getBoolAttribute(i, ['repo-gpgcheck'])
```
?

Or:
```python
getBoolAttribute(getStrAttribute(i, ['repo-gpgcheck'], default='False'))
```
From my POV, it seems easier to read. I don't care about performance here for the last proposal. :p

stormi replied:
Wait a minute. Doesn't getBoolAttribute already handle None as a special default value?

```python
def getBoolAttribute(el, attrs, default = None):
    mandatory = (default == None)
    val = getStrAttribute(el, attrs, '', mandatory).lower()
    if val == '':
        return default
    return val in ['yes', 'true']
```

stormi replied:
Oh, there's this `mandatory` parameter though :thinking:

stormi replied:
The workaround is not pretty, but I don't find either of Ronan's proposals more readable, personnally (does the second one actually work?).

ydirson replied:
Regardless of the logic operator to use (I guess you meant `or` ?), that would cause an explicit `False` on the source (meant to override the default of `True`) to be indistinguishable from `None` (meant to use the default, especially important here when not the overridden by the global setting).

ydirson replied:
> Oh, there's this `mandatory` parameter though thinking

That's the problem, omitting the `default` parameter means making the attribute mandatory, and this is precisely detected using `None`.

But then, it seems pretty sane that `getBoolAttribute()` can only return a bool and never `None`, so I don't think we can do any better, short of introducing an explicit tristate type, which would clearly look like overkill.
